qualtran/bloqs/rotations/hamming_weight_phasing.py

```diff
@@ -0,0 +1,87 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from functools import cached_property
+from typing import Dict, Set, TYPE_CHECKING
+
+import attrs
+
+from qualtran import GateWithRegisters, Signature
+from qualtran.bloqs.arithmetic import HammingWeightCompute
+from qualtran.bloqs.basic_gates import ZPowGate
+
+if TYPE_CHECKING:
+    from qualtran import BloqBuilder, SoquetT
+    from qualtran.resource_counting.bloq_counts import BloqCountT
+
+
+@attrs.frozen
+class HammingWeightPhasing(GateWithRegisters):
+    r"""Applies $Z^{\theta}$ to every qubit of input register of size `bitsize`.
+
+    Hamming weight phasing reduces the number of rotations to be synthesized from $n$ (where
+    $n=bitsize$ is the size of the input register) to $\log_2(n)$ via the following steps:
+        1. Compute hamming weight (HW) of input register in using (at-most) $n-1$ ancillas and
+            Toffoli's in a newly allocated output register if size $\log_2(n)$.
+            $HW|x\rangle -> |x\rangle |\text{HW}(x)\rangle$
+        2. Apply $Z^{2^{k}\theta}$ to the k'th qubit of newly allocated hamming weight register.
+        3. Uncompute the hamming weight register and ancillas allocated in Step-1 with 0 Toffoli
+            cost.
+
+    Overall, for an input register of size $n$, the procedure uses $n - \alpha$ Toffoli's and
+    $n - \alpha + \log_2(n)$ ancilla to reduce $n$ rotation synthesis into $\log_2(n)$  rotation
+    synthesis.
+
+    Args:
+        bitsize: Size of input register to apply `Z ** exponent` to.
+        exponent: The exponent of `Z ** exponent` to be applied to each qubit in the input register.
```

Comment:
I think we should stick with theta (or switch theta to exponent in the docstring.)